### ISSUE TITLE
GraphQL-to-LINQ: change the way, fields are resolved

### ISSUE THREAD

**Issue by Horusiath** [fsprojects/FSharp.Data.GraphQL#78]
See also: #61
### Motivation

Current approach to resolve which fields should take part in the LINQ query very simplistic and may not fit more complex scenarios. Right now we're trying to match fields by GraphQL FieldDef.Name and extract property getter from their resolver quotation. 

This works well for examples like `Define.Field("name", String, fun _ x -> x.Name)` - where defined field name matches resolved property, and resolve function is direct property getter - but lets suppose we need something more advanced, i.e. introducing embedded objects based on database output:
#### Case 1

``` fsharp
Define.Field("orderAddress", Address, fun _ x -> {
    Street = x.OrderStreet
    City = x.OrderCity
    PostalCode = x.OrderPostalCode
    Country = x.OrderCountry
})
```

Current implementation will fail, as it will try to attach field by using straightforward address property of containing object. Another failing example would be:
#### Case 2

``` fsharp
// x.Payment is of type Nullable<decimal>
Define.Field("payment", Nullable Float, fun _ x -> x.Payment |> Option.ofNullable |> Option.map double)
```

This will fail to, as field resolver is not straightforward property getter. This any other cases make GraphQL to LINQ a very limited feature.
### Idea

We could potentially replace current implementation with some kind of _property access analysis_ within GraphQL field resolvers.

How this would work? As each field resolver is F# quotation, we could traverse it and mark all object properties, that are accessed as part of that resolver. For **Case 1**, it's quotation would look like following:

```
Lambda (x,
    NewRecord (Address, PropertyGet (Some (x), Street, []),
               PropertyGet (Some (x), City, []),
               PropertyGet (Some (x), PostalCode, []),
               PropertyGet (Some (x), Country, [])))
```

As we can see traversing that expression tree would allow us to eventually get all 4 properties that are accessed this way. Another example (**Case 2**) will produce following quotation:

```
Lambda (x,
    Call (None, op_PipeRight,
          [Call (None, op_PipeRight,
                 [PropertyGet (Some (x), Payment, []),
                  Lambda (value, Call (None, OfNullable, [value]))]),
           Let (mapping, Lambda (value, Call (None, ToDouble, [value])),
                Lambda (option, Call (None, Map, [mapping, option])))]))
```

While slightly more confusing (thanks to pipe operator), we still can resolve PropertyGet info.
#### Case 3

Third option, we need to think of, is deeply nested property getters:

``` fsharp
<@ fun (outer: Outer) -> outer.X.Y @>
```

Produced expression:

```
Lambda (outer, PropertyGet (Some (PropertyGet (Some (outer), X, [])), Y, []))
```

Of course, this approach will have some limitations i.e.:
- Because it needs to traverse expression tree, it needs to operate on quotation directly or use `[<ReflectedDefinition>]`, nested function calls won't be investigated.
- Point above also can produce hard-to-find issues like structural equality `x = y`, where no all fields of x/y have been fetched. Currently F# compiler automatically assumes some typed don't have null as a valid value (i.e. records, functions), but they actually can be.
##### A comprehensive test suite

Basically every F# quotation Expr will need it test case. I'm already building one, it's designed only to assert if correct properties are retrieved from provided quotation, so we can test this in separate from GraphQL-to-LINQ interpreter itself.
##### Input tracking

We should be able to handle cases such as this:

``` fsharp
<@ fun x ->
    let y = x.Y
    let z = y.Z
@>
```

So we not only need to collect properties based on input variable, but also track all subsequent properties from variables, that originate from it.
